api/routes.go

```diff
@@ -91,53 +115,57 @@ func (api *API) logRequest(r *http.Request) {
 	api.staticLogger.Tracef("Processing request: %v %v, Auth: %v, Skynet Cookie: %v, Referer: %v, Host: %v, RemoreAddr: %v", r.Method, r.URL, hasAuth, hasCookie, r.Referer(), r.Host, r.RemoteAddr)
 }
 
-// tokenFromRequest extracts the JWT token from the request and returns it.
-// It first checks the request headers and then the cookies.
-func tokenFromRequest(r *http.Request) (string, error) {
-	// Check the headers for a token.
-	authHeader := r.Header.Get("Authorization")
-	parts := strings.Split(authHeader, "Bearer")
-	if len(parts) == 2 {
-		return strings.TrimSpace(parts[1]), nil
-	}
-	// Check the cookie for a token.
-	cookie, err := r.Cookie(CookieName)
-	if errors.Contains(err, http.ErrNoCookie) {
-		return "", errors.New("no authorisation token found")
-	}
+// tokenFromAPIKey returns a token, generated for the owner of the given API key.
+func (api *API) tokenFromAPIKey(ctx context.Context, ak string) (jwt2.Token, error) {
+	u, err := api.staticDB.UserByAPIKey(ctx, ak)
 	if err != nil {
-		return "", errors.AddContext(err, "cookie exists but it's not valid")
+		return nil, err
 	}
-	var value string
-	err = secureCookie.Decode(CookieName, cookie.Value, &value)
-	if err != nil {
-		return "", err
-	}
-	return value, nil
+	tk, _, err := jwt.TokenForUser(u.Email, u.Sub)
+	return tk, err
 }
 
-// userFromRequest returns a user object based on the JWT within the request.
-// Note that this method does not rely on a token being stored in the context.
-func (api *API) userFromRequest(r *http.Request) *database.User {
-	t, err := tokenFromRequest(r)
-	if err != nil {
-		return nil
+// apiKeyFromRequest extracts the API key from the request and returns it.
+// It first checks the headers and then the query.
+func apiKeyFromRequest(r *http.Request) (string, error) {
+	// Check the headers for an API key.
+	ak := r.Header.Get(APIKeyHeader)
+	// If there is no API key in the headers, try the query.
+	if ak == "" {
+		ak = r.Form.Get("api_key")
 	}
-	token, err := jwt.ValidateToken(t)
-	if err != nil {
-		return nil
-	}
-	tokenMap, err := token.AsMap(r.Context())
-	if err != nil {
-		return nil
+	if ak == "" {
+		return "", ErrNoAPIKey
 	}
-	sub, exists := tokenMap["sub"]
-	if !exists {
-		return nil
+	return ak, nil
```

Comment:
should we validate the API key here? should we make it its own type to have some extra type safety?

Reply:
Yeah. If the user supplies something we should parse it, validate it and then put it in a type. Having a type will make sure we will never use unvalidated user input by accident.